src/main/java/com/zaxxer/hikari/pool/ProxyConnection.java

```diff
@@ -170,7 +171,7 @@ final void untrackStatement(final Statement statement)
 
    final void markCommitStateDirty()
    {
-      if (isAutoCommit) {
+      if (isAutoCommit || isReadOnly) {
```

Comment:
This is more expensive.  ``markCommitStateDirty()`` may be called dozens or hundreds of times in a single connection use, so ``isReadOnly`` is evaluated many times.  However, ``close()`` is only called once per connection use, and therefore checking the ``isReadOnly`` flag there (where it is currently), is more efficient.